File: services/dataProcessor.js
// Convert NDJSON into array of JSON objects
function processNDJSON(data) {
  const lines = data.trim().split('\n').filter(line => line.trim());

  const jsonData = lines.map(line => JSON.parse(line));

  return jsonData;
}

// Group tick data into time (like 1min, 5min)
function aggregateTicks(ticks, timeframe) {
  const grouped = {};
  ticks.forEach(tick => {
    const timeValue = tick.time || tick.ts;
    const timeKey = getTimeKey(timeValue, timeframe);
    if (!grouped[timeKey]) {
      grouped[timeKey] = [];
    }
    grouped[timeKey].push(tick);
  });

  // Create summary
  const result = Object.entries(grouped).map(([time, group]) => {
    return {
      time: new Date(time),              
      open: group[0].price,             
      high: Math.max(...group.map(t => t.price)), 
      low: Math.min(...group.map(t => t.price)),   
      close: group[group.length - 1].price,       
      volume: group.reduce((sum, t) => sum + t.size, 0), 
      count: group.length               
    };
  });

  return result;
}

// Convert any timestamp into normalized time 
function getTimeKey(date, timeframe) {
  const d = new Date(date);

  if (isNaN(d.getTime())) {
    throw new Error(`Invalid time: ${date}`);
  }

  switch (timeframe) {
    case '1s': 
      d.setMilliseconds(0);
      break;

    case '1m':
      d.setMilliseconds(0);
      d.setSeconds(0);
      break;

    case '5m': 
      d.setMilliseconds(0);
      d.setSeconds(0);
      d.setMinutes(Math.floor(d.getMinutes() / 5) * 5);
      break;
  }

  return d.toISOString();
}

module.exports = {
  processNDJSON,
  aggregateTicks,
  getTimeKey
};

File: pybackend/data_processor.py
from datetime import datetime
from typing import List, Dict, Any


def get_time_key(ts: datetime, timeframe: str) -> datetime:
    d = ts.replace(microsecond=0)
    if timeframe == "1s":
        return d
    if timeframe == "1m":
        return d.replace(second=0)
    if timeframe == "5m":
        minute = (d.minute // 5) * 5
        return d.replace(second=0, minute=minute)
    # default to minute
    return d.replace(second=0)


def aggregate_ticks(ticks: List[Dict[str, Any]], timeframe: str) -> List[Dict[str, Any]]:
    grouped: Dict[datetime, List[Dict[str, Any]]] = {}
    for t in ticks:
        time_val = t.get("time") or t.get("ts")
        if not isinstance(time_val, datetime):
            continue
        key = get_time_key(time_val, timeframe)
        grouped.setdefault(key, []).append(t)

    result: List[Dict[str, Any]] = []
    for key in sorted(grouped.keys()):
        group = grouped[key]
        prices = [float(x["price"]) for x in group]
        sizes = [float(x.get("size", 0.0)) for x in group]
        result.append({
            "time": key,
            "open": prices[0],
            "high": max(prices),
            "low": min(prices),
            "close": prices[-1],
            "volume": sum(sizes),
            "count": len(group),
        })
    return result


